### DISCUSSION TITLE
Psycopg query takes much longer to complete than running the same query in pgAdmin

### DISCUSSION THREAD

**Discussion by FilAleksic** [psycopg/psycopg#910] [Q&A]
Hey I'm running into an issue that I haven't been able to find a solution to online

I'm trying to run the following query that returns about 10-20 rows:
```py
query = sql.SQL("SELECT c_1, c_2, c_3 FROM public.{} WHERE c_4 = %s AND c_5 = %s").format(sql.Identifier(param))
data = (data_1, data_2)

with mydb.db_cursor() as cur:
    cur.execute(query, data)
    for item in cur.fetchall():
        some operation...
```

mydb is set up as follows:
```py
@contextmanager
def db_cursor():
    conn = connect(
        host='localhost',
        port=5432,
        dbname='<db>',
        user='<user>',
        password='<pw>',
        cursor_factory=ServerCursor
    )
    try:
        cur = conn.cursor('curname')
        yield cur
        conn.commit()
    except:
        conn.rollback()
        raise
    finally:
        cur.close()
        conn.close()
```

Timing how long it takes to execute the query and fetch the results with cur.fetchall takes about 4-5 seconds. However, performing the **exact same** query in pgAdmin takes about 200ms. This is obviously a huge difference and is causing some huge bottlenecks to my app.

I am running a server side cursor here as that was recommended to me to try elsewhere. I have done this same test with using the default cursor and the whole process still takes 4-5 seconds. The only difference is that when using the server side cursor the cur.execute is much much faster, but cur.fetchall takes so long you end up at roughly the same time spent overall on the query.

I'm not super tech savvy but as the server side cursor can execute the query with speeds on par as running it in pgAdmin but the fetchall is very slow that there is some issue with how I'm connecting to my database? This confuses me as I'm running the database locally and script locally so there shouldn't be any latency there...?

Note that not every query is much slower than manually running it in pgAdmin. For example the following queries take as long as they would in pgAdmin:
```py
query = "SELECT c_1 FROM some_table WHERE c_2 = %s AND c_3 = 'some_str' ORDER BY RANDOM() LIMIT 1;"
data = (data_1, )
with mydb.db_cursor() as cur:
     cur.execute(query, data)
     out = cur.fetchall()[0][0]

------------------------------------------------------------------

query = sql.SQL("SELECT c_1, c_2, c_3 FROM public.{} WHERE board = %s "
                            "AND c_4 = %s AND c_5 = %s").format(sql.Identifier(param))
data = (data_1, data_2, data_3)
cur.execute(query, data)
rows = cur.fetchall()
```

For reference, the query at the start that caused the issues returns 10-20 rows, whereas this last query returns a little under 100.000 rows finishing in 400ms. 

What could be the issue here?

EDIT:
The query that is causing issues is querying a non-partitioned table with an index on c_1 c_4 and c_5
An interesting thing I observed is that this query takes the expected amount of time (150-200ms) when querying the partition of another table. I have no idea why one would be so much slower than the other. The biggest difference between these 2 is that the partition only has an index on c_4. So no index on c_1 or c_5. Is this somehow causing issues?
Again, do note that these queries are all very fast when ran in pgAdmin so this doesn't seem to be a database performance issue I think?

EDIT2:
Partitioning the database that querying was causing issues has increased performance. Querying is still slower than doing the same queries in pgAdmin though and I still don't quite understand how that could be the case

**Comment by dvarrazzo** [comment]
if you are comparing a server-side cursor in postgres vs. a client-side cursor in pgadmin the comparison cannot be the same and the difference is in postgres. You can test the same in pgadmin by using the DECLARE/FETCH statement.
